modules/gapi/src/streaming/vpl/vpl_legacy_source_engine.cpp

```diff
@@ -0,0 +1,321 @@
+                                  ", size: " << out_buf_size <<
+                                  ", offset: " << out_buf_ptr_offset <<
+                                  ", W: " << surfW <<
+                                  ", H: " << surfH);
+        GAPI_Assert(false && "Invalid offset");
+    }
+
+    std::unique_ptr<mfxFrameSurface1> handle(new mfxFrameSurface1);
+    memset(handle.get(), 0, sizeof(mfxFrameSurface1));
+
+    handle->Info = frameInfo;
+    handle->Data.B = buf + out_buf_ptr_offset;
+    handle->Data.G = handle->Data.B + 1;
+    handle->Data.R = handle->Data.B + 2;
+    handle->Data.A = handle->Data.B + 3;
+    handle->Data.Pitch = surfW;
+
+    return Surface::create_surface(std::move(handle), out_buf_ptr);
+}
+
+surface_ptr_t create_surface_other(mfxFrameInfo frameInfo,
+                                   std::shared_ptr<void> out_buf_ptr,
+                                   size_t out_buf_ptr_offset,
+                                   size_t out_buf_size)
+{
+    mfxU8* buf = reinterpret_cast<mfxU8*>(out_buf_ptr.get());
+    mfxU16 surfH = frameInfo.Height;
+    mfxU16 surfW = (frameInfo.FourCC == MFX_FOURCC_P010) ? frameInfo.Width * 2 : frameInfo.Width;
+    
+    // TODO more intelligent check
+    if (out_buf_size <=
+        out_buf_ptr_offset + (surfW * surfH) + ((surfW / 2) * (surfH / 2))) {
+        GAPI_LOG_WARNING(nullptr, "Not enough buffer, ptr: " << out_buf_ptr <<
+                                  ", size: " << out_buf_size <<
+                                  ", offset: " << out_buf_ptr_offset <<
+                                  ", W: " << surfW <<
+                                  ", H: " << surfH);
+        GAPI_Assert(false && "Invalid offset");
+    }
+
+    std::unique_ptr<mfxFrameSurface1> handle(new mfxFrameSurface1);
+    memset(handle.get(), 0, sizeof(mfxFrameSurface1));
+    
+    handle->Info = frameInfo;
+    handle->Data.Y     = buf + out_buf_ptr_offset;
+    handle->Data.U     = buf + out_buf_ptr_offset + (surfW * surfH);
+    handle->Data.V     = handle->Data.U + ((surfW / 2) * (surfH / 2));
+    handle->Data.Pitch = surfW;
+
+    return Surface::create_surface(std::move(handle), out_buf_ptr);
+}
+
+
+    
+LegacyDecodeSession::LegacyDecodeSession(mfxSession sess,
+                                         DecoderParams&& decoder_param,
+                                         VPLLegacyDecodeEngine::file_ptr&& source) :
+    EngineSession(sess, std::move(decoder_param.stream)),
+    mfx_decoder_param(std::move(decoder_param.param)),
+    source_handle(std::move(source)),
+    procesing_surface_ptr(),
+    output_surface_ptr()
+{
+}
+
+void LegacyDecodeSession::swap_surface(VPLLegacyDecodeEngine& engine) {
+    VPLAccelerationPolicy* acceleration_policy = engine.get_accel();
+    GAPI_Assert(acceleration_policy && "Empty acceleration_policy");
+    auto old_locked = procesing_surface_ptr.lock();
+    try {
+        auto cand = acceleration_policy->get_free_surface(decoder_pool_id).lock();
```

Comment:
Where are the `unlock`s?

Reply:
It is not mutex:)
`get_free_surface` returns `weak_ptr` here. To upgrade it for `shared_ptr` we need `lock()`

this is a seldom place where `auto` is confusing